### PR TITLE
RPC to only shard leaders when quering segments

### PR DESCRIPTION
**Pull request by yah01**
Signed-off-by: yah01 <yang.cen@zilliz.com>
#17411

### CODE REVIEW

File: internal/querycoord/cluster.go [hunk 1]
```diff
@@ -457,46 +426,60 @@ func (c *queryNodeCluster) GetSegmentInfo(ctx context.Context, in *querypb.GetSe
...
-	wg.Add(cnt)
-	for _, node := range c.nodes {
-		go func(node Node) {
-			defer wg.Done()
-			res, err := node.getSegmentInfo(ctx, in)
-			resChan <- respTuple{
-				res: res,
-				err: err,
+	var hasAvailableReplica bool
+	for _, replica := range replicas {
```
Comment (jiaoew1991) [inline internal/querycoord/cluster.go:449]: Serial algorithm performance is slow？

File: internal/querycoord/cluster.go [hunk 2]
```diff
@@ -457,46 +426,60 @@ func (c *queryNodeCluster) GetSegmentInfo(ctx context.Context, in *querypb.GetSe
...
+			break
+		} else { // Abort the partial growing segments
+			segmentInfos = segmentInfos[:sealedSegmentNum]
+		}
+	}
+	c.RUnlock()
+
+	// Return the last error if no available replica
+	if !hasAvailableReplica {
+		return nil, err
```
Comment (jiaoew1991) [inline internal/querycoord/cluster.go:482]: What errors might be returned here？

File: internal/querycoord/cluster.go (hunk 1 above)
Comment (yah01) [reply internal/querycoord/cluster.go:449]: This loop stopped as it saw the first  available replica, and the number of replicas won't be large (<=3)

File: internal/querycoord/cluster.go (hunk 2 above)
Comment (yah01) [reply internal/querycoord/cluster.go:482]: The last RPC error occurred in the loop above (line 459)

**Comment by xiaofan-luan** [comment]
I don't think we should expose GetSegmentInfo to client.
User should actually take care about segment level things. 
Even they care about it, they should get segment info from datacoord rather then querycoord.

**Comment by yah01** [comment]
GetSegmentInfo will be deprecated, close it, #17411 should be also closed after that.

**Pull request closed without merging**